sql/group.go

```diff
@@ -411,14 +416,41 @@ func (v *extractAggregatesVisitor) Visit(expr parser.Expr, pre bool) (parser.Vis
 	return v, expr
 }
 
-func (v *extractAggregatesVisitor) run(n *groupNode, expr parser.Expr) (parser.Expr, error) {
-	*v = extractAggregatesVisitor{n: n}
-	expr = parser.WalkExpr(v, expr)
-	return expr, v.err
+func (n *groupNode) visitor(groupBy parser.GroupBy) *extractAggregatesVisitor {
+	v := &extractAggregatesVisitor{
+		n:         n,
+		groupStrs: make(map[string]struct{}, len(groupBy)),
+	}
+
+	for i := range groupBy {
+		v.groupStrs[groupBy[i].String()] = struct{}{}
+	}
+
+	return v
 }
 
-func (p *planner) extractAggregateFuncs(n *groupNode, expr parser.Expr) (parser.Expr, error) {
-	return p.extractAggregatesVisitor.run(n, expr)
```

Comment:
you should restore something like this; because these visitors end up heap-allocated, they were added to the planner struct so that they could all be allocated simultaneously. This change drops that optimization.